sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala

```diff
@@ -88,24 +85,54 @@ case class AnalyzeTableCommand(tableName: String) extends RunnableCommand {
             }
           }.getOrElse(0L)
 
-        // Update the Hive metastore if the total size of the table is different than the size
-        // recorded in the Hive metastore.
-        // This logic is based on org.apache.hadoop.hive.ql.exec.StatsTask.aggregateStats().
-        if (newTotalSize > 0 && newTotalSize != oldTotalSize) {
-          sessionState.catalog.alterTable(
-            catalogTable.copy(
-              properties = relation.catalogTable.properties +
-                (AnalyzeTableCommand.TOTAL_SIZE_FIELD -> newTotalSize.toString)))
-        }
+        updateTableStats(
+          catalogTable,
+          oldTotalSize = catalogTable.stats.map(_.sizeInBytes.toLong).getOrElse(0L),
+          oldRowCount = catalogTable.stats.flatMap(_.rowCount.map(_.toLong)).getOrElse(-1L),
+          newTotalSize = newTotalSize)
+
+      // data source tables have been converted into LogicalRelations
+      case logicalRel: LogicalRelation if logicalRel.catalogTable.isDefined =>
+        val table = logicalRel.catalogTable.get
+        updateTableStats(
+          table,
+          oldTotalSize = table.stats.map(_.sizeInBytes.toLong).getOrElse(0L),
+          oldRowCount = table.stats.flatMap(_.rowCount.map(_.toLong)).getOrElse(-1L),
+          newTotalSize = logicalRel.relation.sizeInBytes)
 
       case otherRelation =>
-        throw new AnalysisException(s"ANALYZE TABLE is only supported for Hive tables, " +
-          s"but '${tableIdent.unquotedString}' is a ${otherRelation.nodeName}.")
+        throw new AnalysisException(s"ANALYZE TABLE is not supported for " +
+          s"${otherRelation.nodeName}.")
     }
+
+    def updateTableStats(
+        catalogTable: CatalogTable,
+        oldTotalSize: Long,
+        oldRowCount: Long,
+        newTotalSize: Long): Unit = {
+      var newStats: Option[Statistics] = None
+      if (newTotalSize > 0 && newTotalSize != oldTotalSize) {
+        newStats = Some(Statistics(sizeInBytes = newTotalSize))
```

Comment:
@cloud-fan When we have row count info, we can't tell whether the row count is valid or not - we can also run ANALYZE TABLE, and run `INSERT INTO`, then run `ANALYZE TABLE NOSCAN`. So removing it can make sure we only have the right answer.